### ISSUE TITLE
Intellisense no longer resolves overloaded range algorithm with range as first argument?

### ISSUE THREAD

**Issue by jducaud** [microsoft/vscode-cpptools#10252]
### Environment

- Windows 11 Pro 22H2 22621.900
- VS Code 1.74.0
- C/C++ Extension 1.13.6
- WSL:Ubuntu-22.04 with gcc/g++ 12.1 (-std=c++23)
- Single project workspace

Today (2022-12-07 - because yesterday - 2022-12-07 - everything was fine), Intellisense reports multiple errors in the VS Code problems panel. All those errors concern the C++ ranges library:
- the overloaded range algorithms with iterator/sentinel parameters are resolved
- the overloaded range algorithms with range parameter are no longer resolved
- the "pipe composition style" of the range adaptors is no longer supported

Only intellisense is affected. The compilation/run task (in WSL and Compiler Explorer) is performed without errors.

### Bug Summary and Steps to Reproduce

Bug Summary: Intellisense yields false errors regarding some ranges library features:
- the overloaded range algorithms with range parameter are not resolved
- the "pipe composition style" of the range adaptors is not supported

Steps to reproduce:
Try:
```
std::string s{"hello WORLD!"};
std::ranges::transform(s, s.begin(),
                       [](unsigned char c) -> unsigned char
                       { return std::toupper(c); });
```
And
```
for (int i : std::views::iota(1) | std::views::take(7))
    std::cout << i << ' ';
std::cout << std::endl;
```

### Expected behavior

Intellisense should not yield those false errors. It compiles and runs fine with gcc 12.1 in WSL/Ubuntu-22.04 and it compiles and runs fine with x86-64 gcc 12.1 in [Compiler Explorer](https://godbolt.org/z/sK9T4xzeW).

Yesterday everything was working fine and I did not modify any settings in VS Code since.

### Code sample and Logs

```shell
// Code sample ---------------------------------------------
#include <algorithm>
#include <iostream>
#include <ranges>
#include <string>

int main()
{
    std::string s{"hello WORLD!"};
    // the "iterator/sentinel parameters" overload of "__transform_fn" is still resolved
    std::ranges::transform(s.begin(), s.end(), s.begin(),
                           [](unsigned char c) -> unsigned char
                           { return std::tolower(c); });
    // the "range parameter" overload of "__transform_fn" is not resolved anymore
    std::ranges::transform(s, s.begin(),
                           [](unsigned char c) -> unsigned char
                           { return std::toupper(c); });
    std::cout << s << std::endl;

    // the "pipe composition style" is no longer supported
    for (int i : std::views::iota(1) | std::views::take(7))
        std::cout << i << ' ';
    std::cout << std::endl;
    // the "functional composition style" fails because the "range parameter" overload
    // of "_Take" is not resolved anymore
    for (int i : std::views::take(std::views::iota(1), 7))
        std::cout << i << ' ';
    std::cout << std::endl;
}

// c_cpp_properties.json ----------------------------------
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "gnu17",
            "cppStandard": "gnu++23",
            "intelliSenseMode": "linux-gcc-x64",
            "configurationProvider": "ms-vscode.cmake-tools"
        }
    ],
    "version": 4
}
```

### Screenshots

![2022-12-08 - 16-06-41 - Code](https://user-images.githubusercontent.com/55034588/206484025-7d3ce068-5332-4d0b-829f-17fff5aee292.png)
CTRL+F5 in VS Code yields:
```
 *  Executing task in folder test: make 

g++ -std=gnu++23 -Wall -ggdb -Iinclude src/main.cpp -o bin/main 
 *  Terminal will be reused by tasks, press any key to close it.
```
The output is:
```
HELLO WORLD!
1 2 3 4 5 6 7 
1 2 3 4 5 6 7 
[1] + Done                       "/usr/bin/gdb" --interpreter=mi --tty=${DbgTerm} 0<"/tmp/Microsoft-MIEngine-In-5xrkqmu3.55q" 1>"/tmp/Microsoft-MIEngine-Out-mh53pvru.2qs"
```

**Comment by jducaud** [comment]
Note: in the screenshot I simplified the second transform operation by removing "std::toupper", just to see if that would change the outcome. It did not change anything.

**Comment by jducaud** [comment]
I have just installed the "Microsoft Build Tools for Visual Studio 2022" and used them ("cl.exe" with the "/std:c++latest" option to be precise) inside VS Code in Windows (hence no WSL2 here).

VS Code C++ linter correctly handles the ranges library features mentioned above with msvc (cl.exe).

To sum up:
- g++ 12.1 in WSL2 correctly compiles and runs the sample code
- in WSL2-Ubuntu-22.04, VS Code C++ linter displays false errors in the problems panel
- cl.exe 19.34.31935 in Windows correctly compiles and runs the sample code
- in Windows, VS Code C++ linter does not display false errors

This problem seems to be related to the "remote WSL2 thing"...

**Comment by jducaud** [comment]
I rolled back VS Code from 1.74.0 to 1.71.2... The issue was still there. Then I rolled back the C/C++ extension from 1.13.6 to 1.12.4 and found out there was no issue anymore.

Conclusion: The C/C++ extension 1.13.6 brings an issue (not present in the 1.12.4 version) with the WSL2 remote environment.

So I will stay with the 1.12.4 version, until this linter bug is fixed.

**Comment by Gryfenfer97** [comment]
I have the same error with the following code
```cpp
#include <vector>
#include <algorithm>

int main(){
    std::vector<int> v = {1, 2, 3, 4};
    std::ranges::find(v.begin(), v.end(), 1); // No error
    std::ranges::find(v, 1); // "no instance of overloaded function ..."
}
```

the error:
```
no instance of overloaded function "std::ranges::__find_fn::operator()" matches the argument list C/C++(304)
main.cpp(7, 5): argument types are: (std::vector<int, std::allocator<int>>, int)
main.cpp(7, 5): object type is: const std::ranges::__find_fn
```

And my `settings.json`
```json
{
    "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools",
    "C_Cpp.default.cppStandard": "c++20",
    "C_Cpp.default.cStandard": "c17",
    "C_Cpp.default.intelliSenseMode": "linux-gcc-x64"
}
```

And the same thing, everything compile fine but I get this intellisense error message.

**Environment**

- ArchLinux
- VS Code 1.74.0
- C/C++ Extension 1.13.7
- g++ 12.2.0 (c++20)

**Comment by mly32** [comment]
Bumping for visibility on this.

I was able to resolve it for now with @jducaud's solution.

**Comment by jducaud** [comment]
Hello Sean (@sean-mcmanus ), have you any good news about this VS Code internal bug? Is there a way to follow its processing? I tried unsuccessfully to find a VS Code repository issue #1706892 - It seems that the issue numbers are more around 172000 on this repository nowadays, but maybe, as it's an "internal" bug, it has not public visibility... or it is simply a typo in your message? (an unwanted extra-digit) Regards, Jerome.

**Comment by sean-mcmanus** [comment]
@jducaud The bug is in the internal VS bug database so it's not publicly visible -- the bug is active/approved...no other updates yet. I've added a message to the issue.

**Comment by jducaud** [comment]
Thank you @sean-mcmanus !

**Comment by Jesseeee** [comment]
Same issue, keep getting errors on valid std::ranges calls, doesn't matter which function of the ranges library I use

**Comment by sean-mcmanus** [comment]
The fix is available with 1.15.0 (pre-release): https://github.com/microsoft/vscode-cpptools/releases/tag/v1.15.0

If you're still hitting an incorrect IntelliSense error with ranges, you're probably hitting a different root cause bug, such as https://github.com/microsoft/vscode-cpptools/issues/9953 and https://github.com/microsoft/vscode-cpptools/issues/8039, or let us know the details in a new issue if you believe it's a different root cause.
